tensorflow/compiler/xla/service/gpu/miopen_conv_algorithm_picker.cc

```diff
@@ -224,45 +303,57 @@ MiopenConvAlgorithmPicker::PickBestAlgorithm(
+
+  std::vector<MiopenConvAlgorithmPicker::AutotuneResult> profile_results;
+
+  for (const AlgorithmDesc& alg : algorithms) {
+    XLA_SCOPED_LOGGING_TIMER_LEVEL(
+        absl::StrCat("CudnnConvAlgorithmPicker::PickBestAlgorithm algo ",
+                     AlgorithmToString(alg)),
+        2);
+
+    ScratchAllocator scratch_allocator(device_ordinal, allocator);
+    se::dnn::ProfileResult profile_result;
+    VLOG(3) << "Trying algorithm " << AlgorithmToString(alg) << " for "
+            << instr->ToString();
 
-  ScratchAllocator scratch_allocator(device_ordinal, allocator);
-  se::dnn::ProfileResult profile_result;
-  VLOG(3) << "Auto-tuning for " << instr->ToString();
-  RunConvOptions options;
-  options.profile_result = &profile_result;
-  options.first_call_from_algorithm_picker = true;
-  bool launch_ok =
+    // Use assignment instead of brace-list to make GCC 4.9 happy.
+    RunConvOptions options;
+    options.profile_result = &profile_result;
+    options.algo_override = alg;
+    Status launch_status =
         RunCudnnConv(instr, absl::MakeSpan(operand_buffers), result_buffer,
-                     &scratch_allocator, &stream, options)
-            .ok();
+                     &scratch_allocator, &stream, options);
+
+    if (!launch_status.ok()) {
+      continue;
+    }
+
+    if (!profile_result.is_valid()) {
+      continue;
+    }
+
+    profile_results.emplace_back();
+    MiopenConvAlgorithmPicker::AutotuneResult& result = profile_results.back();
+    result.algorithm = alg.algo_id();
+    result.tensor_ops_enabled = alg.tensor_ops_enabled();
 
-  if (launch_ok && profile_result.is_valid()) {
     int64 scratch_bytes_used = scratch_allocator.TotalAllocatedBytes();
-    VLOG(3) << "Auto-tuning succeeded, taking "
-            << profile_result.elapsed_time_in_ms() << "ms and using "
-            << NumBytesToString(scratch_bytes_used)
-            << " of scratch (Best result: " << best_result.elapsed_time_in_ms()
-            << "ms, " << NumBytesToString(best_result_bytes_used)
-            << " of scratch)";
-    best_result = profile_result;
-    best_result_bytes_used = scratch_bytes_used;
-  } else {
-    VLOG(3) << "Auto-tuning failed.";
+    result.scratch_bytes = scratch_bytes_used;
+    result.runtime = absl::Milliseconds(profile_result.elapsed_time_in_ms());
   }
 
-  if (best_result.is_valid()) {
-    VLOG(2) << "Best algorithm for " << instr->ToString() << ": "
-            << AlgorithmToString(best_result.algorithm()) << ", takes "
-            << best_result.elapsed_time_in_ms() << "ms, and uses "
-            << best_result_bytes_used << "B of scratch memory.";
-    return AutotuneResult{best_result.algorithm().algo_id(),
-                          best_result.algorithm().tensor_ops_enabled(),
-                          best_result_bytes_used,
-                          absl::Milliseconds(best_result.elapsed_time_in_ms())};
+  const auto& best_result = absl::c_min_element(
```

Comment:
Is it possible to use the similar decision algorithm [`BestCudnnConvAlgorithm`](https://github.com/tensorflow/tensorflow/blob/3c48e25f067632c1f0f2f54cec6a9a9974e992bc/tensorflow/core/kernels/gpu_utils.cc) as in non-XLA mode? The optimization from line 158-166 in that algorithm make sense to me.